src/names/name-resolution.md

```diff
@@ -1,4 +1,348 @@
+}
+```
+
+```rust,compile_fail,E0659
+pub mod bar {
+    #[macro_export]
+    macro_rules! m {
+        () => {};
+    }
+
+    macro_rules! m2 {
+        () => {};
+    }
+    pub(crate) use m2 as m;
+}
+
+pub fn qux() {
+    use bar::*;
+    m!(); // ERROR: `m` is ambiguous
+}
+```
+
+> [!NOTE]
+> These ambiguity errors are specific to imports, even though they are only observed when those imports are used, having multiple candidates available for a given name during later stages of resolution is not considered an error, so long as none of the imports themselves are ambiguous, there will always be a single unambiguous closest resolution during later stages.
+>
+> ```rust
+> mod bar {
+>     pub struct Name;
+> }
+>
+> mod baz {
+>     pub struct Name;
+> }
+>
+> use baz::Name;
+>
+> pub fn foo() {
+>     use bar::*;
+>     Name; // resolves to bar::Name
+> }
+> ```
+
+r[names.resolution.expansion.imports.ambiguity.moreexpandedvsouter]
+It is an error for name bindings from macro expansions to shadow name bindings from outside of those expansions.
+
+```rust,compile_fail,E0659
+macro_rules! name {
+    () => {}
+}
+
+macro_rules! define_name {
+    () => {
+        macro_rules! name {
+            () => {}
+        }
+    }
+}
+
+fn foo() {
+    define_name!();
+    name!(); // ERROR: `name` is ambiguous
+}
+```
+
+r[names.resolution.expansion.imports.ambiguity.pathvstextualmacro]
+Path-based scope bindings for macros may not shadow textual scope bindings to macros. For bindings from [use declarations], this applies regardless of their [sub-namespace].
+
+```rust,compile_fail,E0659
+#[macro_export]
+macro_rules! m2 {
+    () => {}
+}
+macro_rules! m {
+    () => {}
+}
+pub fn foo() {
+    m!(); // ERROR: `m` is ambiguous
+    use crate::m2 as m; // in scope for entire function body
+}
+```
+
+r[names.resolution.expansion.imports.ambiguity.builtin-attr]
+It is an error to use a user defined attribute or derive macro with the same name as a builtin attribute (e.g. inline).
+
+<!-- ignore: test doesn't support proc-macro -->
+```rust,ignore
+// myinline/src/lib.rs
+use proc_macro::TokenStream;
+
+#[proc_macro_attribute]
+pub fn inline(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    item
+}
+```
+
+<!-- ignore: requires external crates -->
+```rust,ignore
+// src/lib.rs
+use myinline::inline;
+use myinline::inline as myinline;
+
+#[myinline::inline]
+pub fn foo() {}
+
+#[crate::inline]
+pub fn bar() {}
+
+#[myinline]
+pub fn baz() {}
+
+#[inline] // ERROR: `inline` is ambiguous
+pub fn qux() {}
+```
+
+r[names.resolution.expansion.imports.ambiguity.derivehelper]
+Helper attributes may not be used before the macro that introduces them.
+
+> [!NOTE]
+> rustc currently allows derive helpers to be used before their attribute macro introduces them into scope so long as they do not shadow any other attributes or derive helpers that are otherwise correctly in scope. This behavior deprecated and slated for removal.
```

Comment:
```suggestion
> rustc currently allows derive helpers to be used before their attribute macro introduces them into scope so long as they do not shadow any other attributes or derive helpers that are otherwise correctly in scope. This behavior is deprecated and slated for removal.
```